gnucash/gnome-utils/gnc-plugin.c

```diff
@@ -231,155 +223,266 @@ gnc_plugin_get_name (GncPlugin *plugin)
-        gtk_action_set_short_label (action, _(toolbar_labels[i].label));
+        GncDisplayItem *gdi;
+
+        PINFO("action_name is '%s', short_label is '%s'",
+              toolbar_labels[i].action_name, toolbar_labels[i].short_label);
+
+        gdi = (GncDisplayItem*)g_hash_table_lookup (display_item_hash,
+                                                    toolbar_labels[i].action_name);
+        if (gdi)
+            gdi->short_label = toolbar_labels[i].short_label;
     }
 }
 
 
-/** Mark certain actions as "important".  This means that their labels
- *  will appear when the toolbar is set to "Icons and important text"
- *  (e.g. GTK_TOOLBAR_BOTH_HORIZ) mode.
- *
- *  See gnc-plugin.h for documentation on the function arguments. */
+/*  Update the sensitivity of an action */
 void
-gnc_plugin_set_important_actions (GtkActionGroup *action_group,
-                                  const gchar **name)
+gnc_plugin_set_actions_enabled (GSimpleActionGroup *simple_action_group,
+                                const gchar **action_names, gboolean enable)
 {
-    GtkAction *action;
-    gint i;
+    g_return_if_fail (simple_action_group != NULL);
 
-    for (i = 0; name[i]; i++)
+    for (gint i = 0; action_names[i]; i++)
     {
-        action = gtk_action_group_get_action (action_group, name[i]);
-        g_object_set (G_OBJECT(action), "is_important", TRUE, NULL);
+        GAction *action = g_action_map_lookup_action (G_ACTION_MAP(simple_action_group),
+                                                      action_names[i]);
+        if (action)
+             g_simple_action_set_enabled (G_SIMPLE_ACTION(action), enable);
+        else
+            PERR("No such action with name '%s' in action group %p)",
+                  action_names[i], simple_action_group);
     }
-
-    /* If this trips, you've got too many "important" actions.  That
-     * can't *all* be that important, can they? */
-    g_assert(i <= 3);
 }
 
 
-/*  Update a property of existing UI actions.  This function can
- *  modify actions making them visible, invisible, sensitive, or
- *  insensitive.
- *
- *  See gnc-plugin.h for documentation on the function arguments. */
 void
-gnc_plugin_update_actions (GtkActionGroup *action_group,
-                           const gchar **action_names,
-                           const gchar *property_name,
-                           gboolean value)
+gnc_plugin_update_action_labels (GtkWidget *menubar, GtkWidget *toolbar,
+                                 const GncActionUpdate *updates,
+                                 gint n_updates, GtkWidget *statusbar)
 {
-    GtkAction    *action;
-    gint          i;
+    g_return_if_fail (menubar != NULL);
+    g_return_if_fail (toolbar != NULL);
+    g_return_if_fail (updates != NULL || n_updates == 0);
 
-    for (i = 0; action_names[i]; i++)
+    for (gint i = 0; n_updates == -1 ? updates[i].action_name != NULL : i < n_updates; i++)
```

Comment:
Can you explain what's happening here ? The n_updates parameter is never explained, and it seems to have a double use case.

Reply:
OIC, this is factored out from gnc_main_window.cpp and more or less used as is. It's a bit convoluted, but perhaps not worth to improve as part of this PR.

Reply:
I see. The bit I find convoluted is that it has an inline if/then/else clause as the for loop condition. While it works that's pretty hard to read. And it also suggests that in some cases n_updates can be -1 and that that would mean the list of updates is ended with a sentinel entry (an updates record with a NULL action_name). I think that as well is poor design - that you can have two different ways to declare the size of an updates list - either by passing a counter or by setting a NULL value for the last element in the list.
I haven' checked if gnucash has update arrays for both methods. If so, it would be best to rewrite this to use only one method (eitherr all update lists that can be passed to this function have a size counter or they end with a sentinel update value - meaning one that has a NULL entry).

I understand this is not your code originally, but as it's being moved around anyway, it may be worth updating it.

Reply:
I have checked all occurrences and all were being passed the number of elements and none of the arrays had a sentinel so  changed this to always expect the number of elements, I did not notice this as I also used it for the GncDisplayItems.